engines/mohawk/riven_graphics.cpp

```diff
@@ -285,10 +285,17 @@ class TransitionEffectBlend : public TransitionEffect {
 				uint16 *src1 = (uint16 *) _mainScreen->getBasePtr(0, y);
 				uint16 *src2 = (uint16 *) _effectScreen->getBasePtr(0, y);
 				uint16 *dst = (uint16 *) screen->getBasePtr(0, y);
-				for (uint x = 0; x < _mainScreen->w; x++) {
+
+				uint16 widthMulti = _mainScreen->format.bytesPerPixel / 2;
+				for (uint x = 0; x < _mainScreen->w * widthMulti; x++) {
 					uint8 r1, g1, b1, r2, g2, b2;
-					Graphics::colorToRGB< Graphics::ColorMasks<565> >(*src1++, r1, g1, b1);
-					Graphics::colorToRGB< Graphics::ColorMasks<565> >(*src2++, r2, g2, b2);
+					if (_effectScreen->format.bytesPerPixel == 2) {
+						Graphics::colorToRGB< Graphics::ColorMasks<565> >(*src1++, r1, g1, b1);
+						Graphics::colorToRGB< Graphics::ColorMasks<565> >(*src2++, r2, g2, b2);
+					} else {
+						Graphics::colorToRGB< Graphics::ColorMasks<8888> >(*src1++, r1, g1, b1);
```

Comment:
This is not correct because the pixel format obtained from `g_system->getSupportedFormats().front();` could be any 32-bits pixel format, and here we assume ARGB.

To be fair, I'm not sure I want to accept the changes to Riven to make it work with any pixel format. I'm not sure moving the color conversion to the engine fixes anything. Did you test the performance of using hardware accelerated color conversion on the 3DS?

Reply:
I've pretty much abandoned this pull request, as I'm making very good progress on getting ScummVM to switch the system output between RGBA8888 and RGB565 on a per-game basis: BallM4788/scummvm@92fa032e94966e76e616f9115d0ab481c57e9913

To answer your question though, I did test the performance of this pull request on real 3DS hardware, and it runs full-speed with no slowdown. But yeah, no hard feelings on my part if you don't merge this.